venom/rpc/reflect/openapi.py

```diff
@@ -1,6 +1,195 @@
+
+
+def parameters_query(method: Method) -> list:
+    fields = [getattr(method.request, f)
+              for f in method.http_field_locations()[HTTPFieldLocation.QUERY]]
+    return [{**QUERY_PARAMETER, **parameter_common(f)} for f in fields]
+
+
+def parameters_body(method: Method) -> list:
+    body_fields = method.http_field_locations()[HTTPFieldLocation.BODY]
+    if not body_fields:
+        return []
+    fields = {f: getattr(method.request, f) for f in body_fields}
+    if fields == method.request.__fields__:
+        param = dict(
+            name=method.request.__meta__.name,
+            schema=TypeMessage(ref=reference_string(method.response))
+        )
+    else:
+        param = dict(
+            name=method.name + '_body',
+            schema=fields_message(fields.values())
+        )
+    return [{**BODY_PARAMETER, **param}]
+
+
+def to_message(param_data: dict) -> ParameterMessage:
+    return ParameterMessage(
+        **{k: param_data[k] for k in field_names(ParameterMessage) if k in param_data}
+    )
+
+
+def parameters_messages(m: Method) -> list:
+    return list(map(to_message, parameters_body(m) + parameters_path(m) + parameters_query(m)))
+
+
+def response_message(method: Method) -> ResponseMessage:
+    return ResponseMessage(
+        description=method.options.get(DESCRIPTION, ''),
+        schema=FieldsMessage(ref=reference_string(method.response))
+    )
+
+
+def method_message(method: Method) -> MethodMessage:
+    return MethodMessage(
+        produces=['application/json'],  # TODO: other formats
+        responses=ResponsesMessage(default=response_message(method)),
+        parameters=parameters_messages(method)
+    )
+
+
+def methods_messages_map(reflect: Reflect) -> dict:
+    result = defaultdict(dict)
+    for k, group in groupby(reflect.methods, key=lambda a: a.http_path):
+        for m in group:
+            result[k.strip('.')][m.http_method.value.lower()] = method_message(m)
+    return result
+
+
+def fields_message(fields, description='') -> FieldsMessage:
+    params = dict(
+        type='object',
+        properties={
+            v.name: type_message(v) for v in fields
+        }
+    )
+    if description:
+        params[DESCRIPTION] = description
+    return FieldsMessage(**params)
+
+
+def field_types_message(message: Message) -> FieldsMessage:
+    return fields_message(
+        fields(message),
+        description=message.__meta__.get(DESCRIPTION, '')
+    )
+
+
+def messages(reflect: Reflect) -> dict:
+    return {
+        m.__meta__.name: field_types_message(m) for m in reflect.messages
+        if not m.__meta__.name == 'Empty'
+        }
+
+
+def service_collection_name(services: set()) -> str:
+    return '; '.join([s.__meta__.name for s in services])
 
 
 def make_openapi_schema(reflect: Reflect) -> OpenAPISchema:
-    raise NotImplementedError
+    return OpenAPISchema(
+        swagger='2.0',
+        schemes=['http'],
```

Comment:
Better to leave schemes out, per spec. Then it uses the scheme it was accessed by.